src/Xrefcheck/Verify.hs

```diff
@@ -439,11 +461,17 @@ verifyReference
         nestingChange _ = 1
 
     checkReferredFileExists file = do
-      let fileExists = readingSystem $ doesFileExist file
-      let dirExists = readingSystem $ doesDirectoryExist file
-
-      unless (fileExists || dirExists) $
+      unless checkExists $
         throwError (LocalFileDoesNotExist file)
+      where
+        checkExists :: Bool
+        checkExists = any (liftA2 (||) filePathExists dirExists) $ M.keys repoInfo
+
+        filePathExists :: FilePath -> Bool
+        filePathExists = equalFilePath $ expandIndirections file
+
+        dirExists :: FilePath -> Bool
+        dirExists = any filePathExists . scanl (</>) "" . splitDirectories
```

Comment:
The name `dirExists` is a bit misleading...

It seems to suggest `dirExists x` will check if the directory `x` exists.

But in fact, it'll check if any of the parent directories of `x` match the free variable `file`.

Reply:
On a similar note: Instead of retrieving the parent dirs of every single file in `RepoInfo` every single time, what if we collected all directories beforehand?

E.g., change `scanRepo` to return this:

```
data RepoInfo = RepoInfo 
  { riFiles :: Map FilePath (Maybe FileInfo)
  , riDirectories :: Set FilePath
  }
```

I believe this would (1) save a lot of work and (2) make the code's intent much clearer.

Reply:
Fixed. Thanks!